v1.23 blog post: Pod Security Admission Graduates to Beta (#30502)
kubernetes/website: content/en/blog/_posts/2021-12-15-pod-security-admission-beta.md

---
layout: blog
title: 'Pod Security Admission (PSA) Graduates to Beta'
date: 2021-12-15
slug: pod-security-admission-beta
---

**Authors:** Jim Angel (Google), Lachlan Evenson (Microsoft)

> With the release of Kubernetes v1.23, [Pod Security Admission (PSA)](/docs/concepts/security/pod-security-admission/) has now entered beta. PSA is a [built-in](/docs/reference/access-authn-authz/admission-controllers/) admission controller that evaluates pod specifications against a predefined set of [Pod Security Standards](/docs/concepts/security/pod-security-standards/) and determines whether to `admit` or `deny` the pod from running. PSA is the successor to [PodSecurityPolicy](/docs/concepts/policy/pod-security-policy/) which was deprecated in v1.21 and is removed in v1.25. Let's explore how Pod Security Admission works in the hope that cluster administrators and developers alike will use this to enforce secure defaults for their workloads. In this blog we will cover the key concepts of Pod Security Admission along with how to use it. 

Reviewer: nit: 
> will be removed in v1.25

Author: Good callout! I wonder if we should instead update `which was deprecated in v1.21` to `which is deprecated in v1.21` instead, to keep everything present tense (inline with the [style guide](https://kubernetes.io/docs/contribute/style/style-guide/#use-present-tense)).

However, not sure how strict that rule would be in this context (speaking as a blog, in a point of time, where one item is past tense and one is future tense) cc @sftim

Reviewer 2: “_which was deprecated in v1.21_” is how I'd say it conversationally:
- A: “I'm using PodSecurityPolicy”
- B: “Kubernetes deprecated that in v1.21”
- A: “It was deprecated in v1.21? I'd better go and read about alternatives!”
- B: “Here's my rate card, let me know if you'd like help here.”

Author: With that same mindset, you think we should use `s/is removed in v1.25/will be removed in v1.25`?

## Why Pod Security Admission

Pod Security Admission overcomes key shortcomings of Kubernetes' existing, but deprecated, PodSecurityPolicy (PSP) mechanism:

   * Policy authorization model — challenging to deploy with controllers.
   * Risks around switching — a lack of dry-run/audit capabilities made it hard to enable PodSecurityPolicy.
   * Inconsistent and Unbounded API — security constraints were ever-changing which led to the API needing changes.

The shortcomings of PSP made it very difficult to use which led the community to reevaluate whether or not a better implementation could achieve the same goals. One of those goals was to provide an out-of-the-box solution to apply security best practices. Pod Security Admission ships with predefined Pod Security levels that a cluster administrator can configure to meet the desired security posture.

> It's important to note that Pod Security Admission doesn't have complete feature parity with the deprecated PodSecurityPolicy. Specifically, it doesn't have the ability to `mutate` or change Kubernetes resources to auto-remediate a policy violation on behalf of the user. Additionally, it doesn't provide fine-grain control over each allowed field and value within a pod specification or any other Kubernetes resource that you may wish to evaluate. If you need more fine-grained policy control then take a look at the [Gatekeeper](https://github.com/open-policy-agent/gatekeeper) and [other](/docs/concepts/security/pod-security-standards/#faq) projects which support such use cases.

Reviewer: should we link to https://github.com/open-policy-agent/gatekeeper-library instead since that is where we maintain the community-owned policies?

Reviewer 3: Because we aren't mentioning that there are PodSecurityPolicy/PodSecurity equivalent policies that Gatekeeper implements I don't think it's worth digging too deep in this blog and hope that if the reader is interested the Gatekeeper docs would eventually point them at the policy library.

   * Privileged — open and unrestricted
   * Baseline — Covers most common known privilege escalations and provides an easier onboarding
   * Restricted — Highly restricted, covering best practices. May cause compatibility issues

Each of these policy levels define which fields are restricted within a pod specification and the allowed values. Some of the fields restricted by these policies include:
   * `spec.containers[*].ports`
   * `spec.volumes[*].hostPath`
   * `spec.securityContext`
   * `spec.containers[*].securityContext`
    
These policy levels can be applied in two different ways. Either via namespace labels or an AdmissionConfiguration resource. Using namespace labels allows for granular per namespace policy selection whereas using AdmissionConfiguration allows cluster-level defaulting along with exemptions.

### Policy modes

Policies are applied using modes. Here is a list of modes:
   * `enforce` — Any Pods that violate the policy will be rejected
   * `audit` — Pods with violations will be allowed and an audit annotation will be added
   * `warn` — Pods that violate the policy will be allowed and a warning message will be sent back to the user.

> In addition to modes you can also pin the policy to a specific version for example v1.22. Pinning to a specific version allows the behavior to remain consistent as policy changes happen over Kubernetes releases.

Reviewer: should we add something about EOL and support for older k8s release policies?

```shell
kubectl -n kube-system exec kube-apiserver-kind-control-plane -it -- kube-apiserver -h | grep "default enabled ones"
```

The output is similar to this:

```shell
...
      --enable-admission-plugins strings
admission plugins that should be enabled in addition
to default enabled ones (NamespaceLifecycle, LimitRanger,
ServiceAccount, TaintNodesByCondition, PodSecurity, Priority,
DefaultTolerationSeconds, DefaultStorageClass,
StorageObjectInUseProtection, PersistentVolumeClaimResize,
RuntimeClass, CertificateApproval, CertificateSigning,
CertificateSubjectRestriction, DefaultIngressClass,
MutatingAdmissionWebhook, ValidatingAdmissionWebhook,
ResourceQuota).
...
```

`PodSecurity` is listed in the group of default enabled admission plugins.

### Configure Pod Security Admission

Policies are applied to a namespace via labels. These labels are as follows:
   * `pod-security.kubernetes.io/<MODE>: <LEVEL>` (required)
   * `pod-security.kubernetes.io/<MODE>-version: <VERSION>` (*optional*, defaults to latest)

> A specific version can be supplied for each enforcement mode. The version pins the policy to the version that was shipped as part of the Kubernetes release. Pinning to a specific Kubernetes version allows for deterministic policy behavior while allowing flexibility for future updates to Security Policy Standards

Reviewer: nit:
> future updates to Pod Security Standards

It's not recommended to alter control plane / clusters after install, so let's build a new cluster with a default policy on all namespaces.

First, delete the current cluster.

```shell
kind delete cluster
```

Create a Pod Security configuration. 

<!-- TODO: link to the official docs after posted: https://github.com/kubernetes/website/blob/dev-1.23/content/en/docs/tasks/configure-pod-container/enforce-standards-admission-controller.md?plain=1#L25-L55
--->

```shell
cat <<EOF > pod-security.yaml
apiVersion: apiserver.config.k8s.io/v1
kind: AdmissionConfiguration
plugins:
- name: PodSecurity
  configuration:
    apiVersion: pod-security.admission.config.k8s.io/v1beta1
    kind: PodSecurityConfiguration
    defaults:
      enforce: "baseline"
      enforce-version: "latest"
      audit: "baseline"
      audit-version: "latest"
      warn: "restricted"
      warn-version: "latest"
    exemptions:
      # Array of authenticated usernames to exempt.
      usernames: []
      # Array of runtime class names to exempt.
      runtimeClasses: []
      # Array of namespaces to exempt.
      namespaces: [kube-system]
EOF
```

> We now have a default baseline policy, next pass it to the kind configuration to enable the `--admission-control-config-file` API server argument and pass the policy file.

Reviewer 4: ```suggestion
We now have a default baseline policy. Next, pass it to the kind configuration to enable the `--admission-control-config-file` API server argument and pass the policy file.
```

```shell
cat <<EOF > kind-config.yaml
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
- role: control-plane
  kubeadmConfigPatches:
  - |
    kind: ClusterConfiguration
    apiServer:
        # enable admission-control-config flag on the API server
        extraArgs:
          admission-control-config-file: /etc/kubernetes/policies/pod-security.yaml
        # mount new file / directories on the control plane
        extraVolumes:
          - name: policies
            hostPath: /etc/kubernetes/policies
            mountPath: /etc/kubernetes/policies
            readOnly: true
            pathType: "DirectoryOrCreate"
  # mount the local file on the control plane
  extraMounts:
  - hostPath: ./pod-security.yaml
    containerPath: /etc/kubernetes/policies/pod-security.yaml
    readOnly: true
EOF
```

```shell
kind create cluster --image kindest/node:v1.23.0 --config kind-config.yaml
```

Let's look at the default namespace.

```shell
kubectl describe namespace default
```

The output is similar to this:

```
Name:         default
Labels:       kubernetes.io/metadata.name=default
Annotations:  <none>
Status:       Active

No resource quota.

No LimitRange resource.
```

> Let's create a new namespace and see if the labels apply there.

Reviewer: might be good to add a snippet about what happens when the label is removed from the namespace and how that impacts update operations for existing pods

Author: Any suggestions or recommendations? I like the idea; not sure the best way to introduce it without feeling forced / distracting.

cat /etc/kubernetes/policies/pod-security.yaml

# check the api server logs
cat /var/log/containers/kube-apiserver*.log 

# check the api server config
cat /etc/kubernetes/manifests/kube-apiserver.yaml 
```

**UPDATE:** The baseline policy permits `allowPrivilegeEscalation`. While I cannot see the Pod Security default levels of enforcement, they are there. Let's try to provide a manifest that violates the baseline by requesting hostNetwork access.

```shell
cat <<EOF | kubectl -n test-defaults apply -f -
apiVersion: v1
kind: Pod
metadata:
  name: pause-privileged
spec:
  containers:
  - name: pause
    image: k8s.gcr.io/pause
  hostNetwork: true
EOF
```

The output is similar to this:

```shell
Error from server (Forbidden): error when creating "STDIN": pods "pause-privileged" is forbidden: violates PodSecurity "baseline:latest": host namespaces (hostNetwork=true)
```

#### Yes!!! It worked! 🎉🎉🎉

I later found out, another way to check if things are operating as intended is to check the raw API server metrics endpoint.

Run the following command:

```
kubectl get --raw /metrics | grep pod_security_evaluations_total
```

The output is similar to this:

```

> # HELP pod_security_evaluations_total [ALPHA] Number of policy evaluations that occurred, not counting ignored or exempt requests.

Reviewer: Is this alpha specifically for the `pod_security_evaluations_total ` metric?

Author: cc @liggitt who shared the `metrics` pro-tip with me.

Reviewer 5: yes, the metric is alpha (metrics currently only have ALPHA or STABLE stability levels)

   * Privileged — open and unrestricted
   * Baseline — Covers most common known privilege escalations and provides an easier onboarding
   * Restricted — Highly restricted, hardening against known and unknown privilege escalations. May cause compatibility issues

Each of these policy levels define which fields are restricted within a pod specification and the allowed values. Some of the fields restricted by these policies include:
   * `spec.securityContext`
   * `spec.hostNetwork`
   * `spec.volumes[*].hostPath`
   * `spec.containers[*].securityContext`

Policy levels are applied via labels on Namespace resources, which allows for granular per-namespace policy selection. The AdmissionConfiguration in the apiserver can also be configured to set cluster-wide default levels and exemptions.

### Policy modes

Policies are applied in a specific mode. Multiple modes (with different policies) can be set on the same namespace. Here is a list of modes:
   * `enforce` — Any Pods that violate the policy will be rejected
   * `audit` — Violations will be recorded as an annotation in the audit logs, but don't affect whether the pod is allowed.

>    * `warn` — Violations will send a warning message back to the user, but don't affect whether the pod is allowed.sent back to the user.

Reviewer 4: ```suggestion
   * `warn` — Violations will send a warning message back to the user, but don't affect whether the pod is allowed.
```

In addition to modes you can also pin the policy to a specific version for example v1.22. Pinning to a specific version allows the behavior to remain consistent as the policy definition changes over Kubernetes releases.

## Hands on demo

### Prerequisites

- [kind CLI installed](https://kind.sigs.k8s.io/docs/user/quick-start/#installation)
- [Docker](https://docs.docker.com/get-docker/) or [Podman](https://podman.io/getting-started/installation) container runtime & CLI

### Deploy a kind cluster

```shell
kind create cluster --image kindest/node:v1.23.0
```

> It might take awhile to start and once it's started it might take a minute or so before the node becomes ready.

Reviewer 4: ```suggestion
It might take a while to start and once it's started it might take a minute or so before the node becomes ready.
```

to default enabled ones (NamespaceLifecycle, LimitRanger,
ServiceAccount, TaintNodesByCondition, PodSecurity, Priority,
DefaultTolerationSeconds, DefaultStorageClass,
StorageObjectInUseProtection, PersistentVolumeClaimResize,
RuntimeClass, CertificateApproval, CertificateSigning,
CertificateSubjectRestriction, DefaultIngressClass,
MutatingAdmissionWebhook, ValidatingAdmissionWebhook,
ResourceQuota).
...
```

`PodSecurity` is listed in the group of default enabled admission plugins.

If using a cloud provider or if you don't have access to the API server, the best way to check would be to run a quick end-to-end test:

```
kubectl create namespace verify-pod-security
kubectl label namespace verify-pod-security pod-security.kubernetes.io/enforce=restricted
kubectl run test --dry-run=server --image=k8s.gcr.io/pause --privileged
kubectl delete namespace verify-pod-security
```

### Configure Pod Security Admission

Policies are applied to a namespace via labels. These labels are as follows:
   * `pod-security.kubernetes.io/<MODE>: <LEVEL>` (required to enable pod security)
   * `pod-security.kubernetes.io/<MODE>-version: <VERSION>` (*optional*, defaults to latest)

> A specific version can be supplied for each enforcement mode. The version pins the policy to the version that was shipped as part of the Kubernetes release. Pinning to a specific Kubernetes version allows for deterministic policy behavior while allowing flexibility for future updates to Security Policy Standards From earlier, recall that the possible modes are `enforce`, `audit` and `warn`.

Reviewer 4: The last part seems disconnected from the paragraph.

> When to use `warn`?
> 
> Don't use `warn` for the exact same level+version of the policy as `enforce`. In the admission sequence, if `enforce` fails, the entire sequence fails before evaluating the `warn`. The typical uses for `warn` are:

Reviewer 4: ```suggestion
When _not_ to use `warn`?

Don't use `warn` for the exact same level+version of the policy as `enforce`. In the admission sequence, if `enforce` fails, the entire sequence fails before evaluating the `warn`. 

When to use `warn`?
The typical uses for `warn` are:
```

* `warn` at the same level but a different version (e.g. pin `enforce` to *restricted+v1.23* and `warn` at *restricted+latest*)
* `warn` at a stricter level (e.g. `enforce` baseline, `warn` restricted)

First, create a namespace called `verify-pod-security` if not created earlier. For the demo, `--overwrite` is used when labeling to allow repurposing a single namespace for multiple examples.

```shell
kubectl create namespace verify-pod-security
```

### Deploy demo workloads

Each workload represents a higher level of security that would not pass the profile that comes after it.

For the following examples, use the `k8s.gcr.io/pause` container which is a small container that runs a never ending `sleep` command. Pod Security is not interested in the container, but rather the Kubernetes manifests that are being admitted.

**Privileged level and workload**

> For the privileged pod, allow [privilege escalation](/docs/concepts/security/pod-security-standards/#privileged). This allows the process inside a container to gain new processes and can be dangerous if untrusted.

Reviewer 4: ```suggestion
For the privileged pod, use the [privileged policy](/docs/concepts/security/pod-security-standards/#privileged). This allows the process inside a container to gain new processes (aka privilege escalation) and can be dangerous if untrusted.
```

First, let's apply a restricted Pod Security level for a test.

```shell
# enforces a "restricted" security policy and audits on restricted
kubectl label --overwrite ns verify-pod-security \
  pod-security.kubernetes.io/enforce=restricted \
  pod-security.kubernetes.io/audit=restricted
```

> Next, try to deploy a provileged workload in the namespace.

Reviewer 4: ```suggestion
Next, try to deploy a privileged workload in the namespace.
```

```

Now let's apply the privileged Pod Security level and try again.

```shell
# enforces a "privileged" security policy and warns / audits on baseline
kubectl label --overwrite ns verify-pod-security \
  pod-security.kubernetes.io/enforce=privileged \
  pod-security.kubernetes.io/warn=baseline \
  pod-security.kubernetes.io/audit=baseline
```

```shell
cat <<EOF | kubectl -n verify-pod-security apply -f -
apiVersion: v1
kind: Pod
metadata:
  name: pause-privileged
spec:
  containers:
  - name: pause
    image: k8s.gcr.io/pause
    securityContext:
      allowPrivilegeEscalation: true
EOF
```

The output is similar to this:

```
pod/pause-privileged created
```

We can run `kubectl -n verify-pod-security get pods` to verify it is running.

**Baseline level and workload**

> The [baseline](/docs/concepts/security/pod-security-standards/#baseline) pod demonstrates sensible defaults but does not allow privilege escalation, so the pod above would fail here too. Baseline accepts null / no values for unset keys.

Reviewer 4: ```suggestion
The [baseline policy](/docs/concepts/security/pod-security-standards/#baseline) demonstrates sensible defaults but does not allow privilege escalation, so the pod above would fail here too. Baseline accepts null / no values for unset keys.
```

```

```shell
cat <<EOF | kubectl -n verify-pod-security apply -f -
apiVersion: v1
kind: Pod
metadata:
  name: pause-baseline
spec:
  containers:
  - name: pause
    image: k8s.gcr.io/pause
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        add:
          - NET_BIND_SERVICE
          - CHOWN
EOF
```

The output is similar the following. Note that the warnings match the error message from the test above, but the pod is still successfully created.

```
Warning: would violate PodSecurity "restricted:latest": unrestricted capabilities (container "pause" must set securityContext.capabilities.drop=["ALL"]; container "pause" must not include "CHOWN" in securityContext.capabilities.add), runAsNonRoot != true (pod or container "pause" must set securityContext.runAsNonRoot=true), seccompProfile (pod or container "pause" must set securityContext.seccompProfile.type to "RuntimeDefault" or "Localhost")
pod/pause-baseline created
```

Remember, we set the `verify-pod-security` namespace to `warn` based on the restricted profile. We can run `kubectl -n verify-pod-security get pods` to verify it is running.

**Restricted level and workload**

> [Restricted](/docs/concepts/security/pod-security-standards/#restricted) requires rejection of all privileged parameters. It is the most secure with a trade-off for complexity. The restricted policy allows containers to add the `NET_BIND_SERVICE` capability.

Reviewer 4: ```suggestion
The [restricted policy](/docs/concepts/security/pod-security-standards/#restricted) requires rejection of all privileged parameters. It is the most secure with a trade-off for complexity. The restricted policy allows containers to add the `NET_BIND_SERVICE` capability.
```

      usernames: []
      # Array of runtime class names to exempt.
      runtimeClasses: []
      # Array of namespaces to exempt.
      namespaces: [kube-system]
EOF
```

We now have a default baseline policy, next pass it to the kind configuration to enable the `--admission-control-config-file` API server argument and pass the policy file.

```shell
cat <<EOF > kind-config.yaml
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
- role: control-plane
  kubeadmConfigPatches:
  - |
    kind: ClusterConfiguration
    apiServer:
        # enable admission-control-config flag on the API server
        extraArgs:
          admission-control-config-file: /etc/kubernetes/policies/pod-security.yaml
        # mount new file / directories on the control plane
        extraVolumes:
          - name: policies
            hostPath: /etc/kubernetes/policies
            mountPath: /etc/kubernetes/policies
            readOnly: true
            pathType: "DirectoryOrCreate"
  # mount the local file on the control plane
  extraMounts:
  - hostPath: ./pod-security.yaml
    containerPath: /etc/kubernetes/policies/pod-security.yaml
    readOnly: true
EOF
```

> 

Reviewer 4: ```suggestion

Create a new cluster using the kind configuration file defined above.

```

pod_security_evaluations_total{decision="allow",mode="enforce",policy_level="privileged",policy_version="latest",request_operation="create",resource="pod",subresource=""} 0
pod_security_evaluations_total{decision="allow",mode="enforce",policy_level="privileged",policy_version="latest",request_operation="update",resource="pod",subresource=""} 0
pod_security_evaluations_total{decision="deny",mode="audit",policy_level="baseline",policy_version="latest",request_operation="create",resource="pod",subresource=""} 1
pod_security_evaluations_total{decision="deny",mode="enforce",policy_level="baseline",policy_version="latest",request_operation="create",resource="pod",subresource=""} 1
pod_security_evaluations_total{decision="deny",mode="warn",policy_level="restricted",policy_version="latest",request_operation="create",resource="controller",subresource=""} 2
pod_security_evaluations_total{decision="deny",mode="warn",policy_level="restricted",policy_version="latest",request_operation="create",resource="pod",subresource=""} 2
```

A monitoring tool could ingest these metrics too for reporting, assessments, or measuring trends.

> ### Auditing

Author: ```suggestion
## Auditing
```

```

Auditing is also a good first step in evaluating your cluster's current compliance with Pod Security. The Kubernetes Enhancement Proposal (KEP) hints at a future where `baseline` [could be the default for unlabeled namespaces](https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/2579-psp-replacement/README.md#rollout-of-baseline-by-default-for-unlabeled-namespaces).

Example `audit-policy.yaml` configuration tuned for Pod Security events:

```
apiVersion: audit.k8s.io/v1
kind: Policy
rules:
- level: RequestResponse
  resources:
    - group: "" # core API group
      resources: ["pods", "deployments", "statefulsets", "daemonsets"]
  verbs: ["create"]
  omitStages:
    - "RequestReceived"
    - "ResponseStarted"
    - "Panic"

> ```

Author: Is there a better policy / filter to use here (for simplicity)?
cc @tallclair @liggitt

Author: + @kubernetes/sig-security-pr-reviews

Once auditing is enabled, look at the configured local file if using `--audit-log-path` or the destination of a webhook if using `--audit-webhook-config-file`.

If using a file (`--audit-log-path`), run `cat /PATH/TO/API-AUDIT/LOG.log | grep "is forbidden:"` to see all rejected workloads audited.

> # PSP migrations

Author: ```suggestion
## PSP migrations
```

## Why Pod Security

Pod Security overcomes key shortcomings of Kubernetes' existing, but deprecated, PodSecurityPolicy (PSP) mechanism:

   * Policy authorization model — challenging to deploy with controllers.
   * Risks around switching — a lack of dry-run/audit capabilities made it hard to enable PodSecurityPolicy.
   * Inconsistent and Unbounded API — the large configuration surface and evolving constraints led to a complex and confusing API.

The shortcomings of PSP made it very difficult to use which led the community to reevaluate whether or not a better implementation could achieve the same goals. One of those goals was to provide an out-of-the-box solution to apply security best practices. Pod Security ships with predefined Pod Security levels that a cluster administrator can configure to meet the desired security posture.

> It's important to note that Pod Security doesn't have complete feature parity with the deprecated PodSecurityPolicy. Specifically, it doesn't have the ability to `mutate` or change Kubernetes resources to auto-remediate a policy violation on behalf of the user. Additionally, it doesn't provide fine-grain control over each allowed field and value within a pod specification or any other Kubernetes resource that you may wish to evaluate. If you need more fine-grained policy control then take a look at the [Gatekeeper](https://github.com/open-policy-agent/gatekeeper) and [other](/docs/concepts/security/pod-security-standards/#faq) projects which support such use cases.

Reviewer 5: nit: not sure why mutate is quoted with back-ticks... it's not really a keyword / code verb

Reviewer 5: nit: make fine-grain and fine-grained consistent?

Reviewer 5: we were pretty careful not to pick just one external policy project when linking to alternatives in https://kubernetes.io/docs/concepts/security/pod-security-standards/#what-s-the-difference-between-a-security-profile-and-a-security-context ... not sure if you want to be similarly noncommittal here

    
Pod Security is a built-in [admission controller](/docs/reference/access-authn-authz/admission-controllers/) starting with Kubernetes v1.22, but can also be run as a standalone [webhook](/docs/reference/access-authn-authz/extensible-admission-controllers/). Admission controllers function by intercepting requests in the Kubernetes API server prior to persistence to storage. They can either `admit` or `deny` a request. In the case of Pod Security, pod specifications will be evaluated against a configured policy in the form of a Pod Security Standard. This means that security sensitive fields in a pod specification will only be allowed to have [specific](h/docs/concepts/security/pod-security-standards/#profile-details) values. 

## Configuring Pod Security
    
### Pod Security Standards
    
In order to use Pod Security we first need to understand [Pod Security Standards](/docs/concepts/security/pod-security-standards/). These standards define three different policy levels that range from permissive to restrictive. These levels are as follows:
   * Privileged — open and unrestricted
   * Baseline — Covers known privilege escalations while minimizing restrictions

>    * Restricted — Highly restricted, hardening against known and unknown privilege escalations. May cause compatibility issues

Reviewer 5: I might suggest making `privileged`, `baseline`, and `restricted` lower-case and code-quoted, since these are keywords in the PodSecurity labeling scheme

Each of these policy levels define which fields are restricted within a pod specification and the allowed values. Some of the fields restricted by these policies include:

>    * `spec.securityContext`

Reviewer 5: the securityContext field itself isn't restricted, but specific subfields are, like `spec.securityContext.sysctls`

   * `spec.hostNetwork`
   * `spec.volumes[*].hostPath`

>    * `spec.containers[*].securityContext`

Reviewer 5: the securityContext field itself isn't restricted, but specific subfields are, like `spec.containers[*].securityContext.privileged`

> Policy levels are applied via labels on Namespace resources, which allows for granular per-namespace policy selection. The AdmissionConfiguration in the api server can also be configured to set cluster-wide default levels and exemptions.

Reviewer 5: nit: api → API

### Policy modes

> Policies are applied in a specific mode. Multiple modes (with different policies) can be set on the same namespace. Here is a list of modes:

Reviewer 5: ```suggestion
Policies are applied in a specific mode. Multiple modes (with different policy levels) can be set on the same namespace. Here is a list of modes:
```

   * `enforce` — Any Pods that violate the policy will be rejected
   * `audit` — Violations will be recorded as an annotation in the audit logs, but don't affect whether the pod is allowed.
   * `warn` — Violations will send a warning message back to the user, but don't affect whether the pod is allowed.

> In addition to modes you can also pin the policy to a specific version for example v1.22. Pinning to a specific version allows the behavior to remain consistent as the policy definition changes over Kubernetes releases. If pinning to a specific Pod Security Standard version in a Kubernetes release you should take the time to understand the Kubernetes [supported version policy](/releases/version-skew-policy/#supported-versions).

Reviewer 5: `... (for example, v1.22).`

Reviewer 5: `Pinning to a specific version allows the behavior to remain consistent if the policy definition changes in future Kubernetes releases.`

Reviewer 5: > ... you should take the time to understand the Kubernetes [supported version policy](/releases/version-skew-policy/#supported-versions).

I'm not sure that's particularly relevant... PodSecurity will continue to honor specific versions of the policy indefinitely.

Reviewer 3: Thanks Jordan - the intent of this statement was if there was a specific bug in a Pod Security Standard that needed fixing and you were using a pinned version you may be out of luck depending on support. Not sure if that would help anyway because I'm not sure you can pin to a patch release (eg. v1.22.4)

Reviewer 3: Yanking that sentence

## Hands on demo

### Prerequisites

- [kind CLI installed](https://kind.sigs.k8s.io/docs/user/quick-start/#installation)
- [Docker](https://docs.docker.com/get-docker/) or [Podman](https://podman.io/getting-started/installation) container runtime & CLI

### Deploy a kind cluster

```shell

> kind create cluster --image kindest/node:v1.23.0

Author: I had the same issue and built one too (then updated PR not to use my local tag). I have it listed in my "steps before hold can be lifted" to include changing to the latest `sha256` tag. With the PR targeting `main` it can merge at anytime, even post-release.

```

### Deploy demo workloads

Each workload represents a higher level of security that would not pass the profile that comes after it.

For the following examples, use the `busybox` container runs a `sleep` command for 1 million seconds (~11 days) or until deleted. Pod Security is not interested in the container, but rather the Kubernetes manifests and corresponding security values that are being admitted.

### Privileged level and workload

For the privileged pod, use the [privileged policy](/docs/concepts/security/pod-security-standards/#privileged). This allows the process inside a container to gain new processes (also known as "privilege escalation") and can be dangerous if untrusted.

First, let's apply a restricted Pod Security level for a test.

```shell
# enforces a "restricted" security policy and audits on restricted
kubectl label --overwrite ns verify-pod-security \
  pod-security.kubernetes.io/enforce=restricted \
  pod-security.kubernetes.io/audit=restricted
```

Next, try to deploy a privileged workload in the namespace.

```shell
cat <<EOF | kubectl -n verify-pod-security apply -f -
apiVersion: v1
kind: Pod
metadata:
  name: busybox-privileged
spec:
  containers:

>   - name: busybox

Reviewer 6: Note to myself: Tutorial is using `nginx` whereas blog is using `busybox`. Switch over the tutorial to use `busybox`

Author: I started using `nginx` and the comment came up that we should use something community owned, like `k8s.gcr.io/pause`, but I ran into a couple snags getting the demo to work smoothly.

I switched back to `nginx`, then later ran into issues when I was applying to the restricted profile (e.g: nginx is built / runs as root, so in my demo, I could either use `nginxinc/nginx-unprivileged` (not preferred) or something less opinionated about the runAsUser so I went with `busybox` + `sleep`. I'm not committed to using `busybox` if there's a compelling reason to use something else!

Reviewer 6: Oh thanks for the history on why you switched to `busybox` from `nginx`. After reading that, makes more sense to continue to use `busybox`!!

```

To solve this, set the effective UID (`runAsUser`) to a non-zero (root) value or use the `nobody` UID (65534).

```shell
# delete the original pod
kubectl -n verify-pod-security delete pod busybox-restricted

# create the pod again with new runAsUser
cat <<EOF | kubectl -n verify-pod-security apply -f -
apiVersion: v1
kind: Pod
metadata:
  name: busybox-restricted
spec:
  securityContext:
    runAsUser: 65534
  containers:
  - name: busybox
    image: busybox
    args:
    - sleep
    - "1000000"
    securityContext:
      seccompProfile:
        type: RuntimeDefault
      runAsNonRoot: true
      allowPrivilegeEscalation: false
      capabilities:
        drop:
          - ALL
        add:
          - NET_BIND_SERVICE
EOF
```

Run `kubectl -n verify-pod-security get pods` to verify it is running. The output is similar to this:

```
NAME                 READY   STATUS    RESTARTS   AGE
busybox-restricted   1/1     Running   0          25s
```

Clean up the demo (restricted pod and namespace) with:

> ```shell

Reviewer 6: Note to Self: Add a clean up section for tutorial

Reviewer 6: Tutorial now has a clean up section

kubectl delete namespace verify-pod-security
```

At this point, if you wanted to dive deeper into linux permissions or what is permitted for a certain container, exec into the control plane and play around with `containerd` and `crictl inspect`.

```shell
# if using docker, shell into the control plane
docker exec -it kind-control-plane bash

# list running containers
crictl ps

# inspect each one by container ID
crictl inspect <CONTAINER ID>
```

### Applying a cluster-wide policy

In addition to applying labels to namespaces to configure policy you can also configure cluster-wide policies and exemptions using the AdmissionConfiguration resource. 

Using this resource, policy definitions are applied cluster-wide by default and any policy that is applied via namespace labels will take precedence. 

> There is no runtime configurable API for this resource and a cluster administrator would need to specify a path to the file below via the `--admission-control-config-file` flag on the API server. 

Reviewer 5: I wouldn't refer to the configuration type as a "resource"... that's typically reserved for URLs of REST APIs.

I'd probably call it "an `AdmissionConfiguration` configuration file."

> In the following resource we are enforcing the baseline policy and warning and auditing the baseline policy. We are also making the kube-system namespace exempt from this policy.

Reviewer 6: ```suggestion
In the following resource we are enforcing the baseline policy and warning and auditing the restricted policy. We are also making the kube-system namespace exempt from this policy.
```

> It's not recommended to alter control plane / clusters after install, so let's build a new cluster with a default policy on all namespaces.

Reviewer 5: on the other hand, this might be a chance to show up the safe rollout capabilities of this feature... you can tighten default warn/audit levels on an existing cluster without breaking writes to existing pods or workloads.

First, delete the current cluster.

```shell
kind delete cluster
```

Create a Pod Security configuration. 

<!-- TODO: link to the official docs after posted: https://github.com/kubernetes/website/blob/dev-1.23/content/en/docs/tasks/configure-pod-container/enforce-standards-admission-controller.md?plain=1#L25-L55
--->

```shell
cat <<EOF > pod-security.yaml
apiVersion: apiserver.config.k8s.io/v1
kind: AdmissionConfiguration
plugins:
- name: PodSecurity
  configuration:
    apiVersion: pod-security.admission.config.k8s.io/v1beta1
    kind: PodSecurityConfiguration
    defaults:
      enforce: "baseline"
      enforce-version: "latest"

>       audit: "baseline"

Reviewer 6: ```suggestion
      audit: "restricted"
```

Author: I'm ok with this suggestion, but curious about your reasoning why? I don't care much if it's `baseline` or `restricted` for auditing, but I approached it as a cluster operator of, "I care historically what's allowed or rejected by PSA" vs. auditing a theoretical higher level of restrictions. I could see value in this approach though if you were interested in, "How many workloads do I allow today that would fail a more restrictive level?" but maybe I would use `warn` for that.

I see the modes like:
- `warn` == for developers / CD signal
- `audit` == for operators / compliance

No matter what approach is used, it might be worth talking about "why;" kind of similar to how we talk about when to `warn`. Wdyt?

Reviewer 6: To be honest, I was only trying to align the label here with the text above and the tutorial for consistency :D 

In terms of personas though, you make a good point. I was mostly thinking about the latter when selecting these levels for audit i.e. "How many workloads do I allow today that would fail a more restrictive level?". But the other scenario is valid too. 

As long as we pick one consistently and share the rationale behind it prior to the `yaml`, Happy with either of them.

Author: I think it's worth adding a blurb about why the levels are chosen, similar to the the "when to warn?" section. I'll probably keep it as is with `baseline` though. Maybe you could do the same for using `restricted` in the tutorial? Open to alternative suggestions too!

Reviewer 6: yes that works!

```shell
cat <<EOF > kind-config.yaml
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
- role: control-plane
  kubeadmConfigPatches:
  - |
    kind: ClusterConfiguration
    apiServer:
        # enable admission-control-config flag on the API server
        extraArgs:
          admission-control-config-file: /etc/kubernetes/policies/pod-security.yaml
        # mount new file / directories on the control plane
        extraVolumes:
          - name: policies
            hostPath: /etc/kubernetes/policies
            mountPath: /etc/kubernetes/policies
            readOnly: true
            pathType: "DirectoryOrCreate"
  # mount the local file on the control plane
  extraMounts:

>   - hostPath: ./pod-security.yaml

Reviewer 6: Maybe use `/tmp` directory for consistency with tutorial: https://github.com/kubernetes/website/blob/004e9a48b3cb7912c85a24f1fcb6812f91abb688/content/en/docs/tutorials/pod-security/cluster-level-pss.md#configure-the-api-server and easier clean up?

Author: IIRC `/tmp` works but, given the permissions that /tmp usually has, you're mounting a very open dir into the containers for potential abuse. While I think it's unlikely, it's probably not a great practice if someone adopts it in production. I like the idea of taking the current directory for the users interpretation.

I don't think either of the tutorials need to change but I could be sold on unifying. (maybe it's a good idea to have diverging examples?)

pod_security_evaluations_total{decision="allow",mode="enforce",policy_level="baseline",policy_version="latest",request_operation="create",resource="pod",subresource=""} 2
pod_security_evaluations_total{decision="allow",mode="enforce",policy_level="privileged",policy_version="latest",request_operation="create",resource="pod",subresource=""} 0
pod_security_evaluations_total{decision="allow",mode="enforce",policy_level="privileged",policy_version="latest",request_operation="update",resource="pod",subresource=""} 0
pod_security_evaluations_total{decision="deny",mode="audit",policy_level="baseline",policy_version="latest",request_operation="create",resource="pod",subresource=""} 1
pod_security_evaluations_total{decision="deny",mode="enforce",policy_level="baseline",policy_version="latest",request_operation="create",resource="pod",subresource=""} 1
pod_security_evaluations_total{decision="deny",mode="warn",policy_level="restricted",policy_version="latest",request_operation="create",resource="controller",subresource=""} 2
pod_security_evaluations_total{decision="deny",mode="warn",policy_level="restricted",policy_version="latest",request_operation="create",resource="pod",subresource=""} 2

> ```

Reviewer 6: This is my favorite section of the blog!!

  verbs: ["create"]
  omitStages:
    - "RequestReceived"
    - "ResponseStarted"
    - "Panic"
```

Once auditing is enabled, look at the configured local file if using `--audit-log-path` or the destination of a webhook if using `--audit-webhook-config-file`.

If using a file (`--audit-log-path`), run `cat /PATH/TO/API-AUDIT/LOG.log | grep "is forbidden:"` to see all rejected workloads audited.

## PSP migrations

If you're already using PSP, SIG Auth has created a guide and [published the steps to migrate off of PSP](/docs/tasks/configure-pod-container/migrate-from-psp/).

To summarize the process:
- Update all existing PSPs to be non-mutating
- Apply Pod Security policies in `warn` or `audit` mode
- Upgrade Pod Security policies to `enforce` mode
- Remove `PodSecurityPolicy` from `--enable-admission-plugins`

Listed as "optional future extensions" and currently out of scope, SIG Auth has kicked around the idea of providing a tool to assist with migrations. More [details in the KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/2579-psp-replacement/README.md#automated-psp-migration-tooling).

## Wrap up

> Pod Security is a promising new feature that provides an out-of-the-box way to allow users to improve the security posture of their workloads. Like any new enhancement that has matured to beta, we ask that you try it out, provide feedback, or share your experience via either raising a Github issue or joining SIG Auth community meetings. It's our hope that Pod Security will be deployed on every cluster in our ongoing pursuit as a community to make Kubernetes security a priority.

Reviewer 6: ```suggestion
Pod Security is a promising new feature that provides an out-of-the-box way to allow users to improve the security posture of their workloads. Like any new enhancement that has matured to beta, we ask that you try it out, provide feedback, or share your experience via either raising a Github issue or joining SIG Auth community meetings. It's our hope that Pod Security will be deployed on every cluster in our ongoing pursuit as a community to make Kubernetes security a priority.
For a step by step guide on how to enable "baseline" Pod Security Standards with Pod Security Admission feature please refer to these dedicated [tutorials](docs/tutorials/pod-security) that cover the configuration needed at cluster level and namespace level. 
```

Reviewer 6: Felt like highlighting the tutorial as part of the wrap up would make it easier to not miss them v/s clubbing them together with other references. Open to other ideas too

---
layout: blog
title: 'Pod Security Graduates to Beta'

> date: 2021-12-15

Reviewer 7: ```suggestion
date: 2021-12-09
```

slug: pod-security-admission-beta
---

**Authors:** Jim Angel (Google), Lachlan Evenson (Microsoft)

> With the release of Kubernetes v1.23, [Pod Security admission](/docs/concepts/security/pod-security-admission/) has now entered beta. Pod Security is a [built-in](/docs/reference/access-authn-authz/admission-controllers/) admission controller that evaluates pod specifications against a predefined set of [Pod Security Standards](/docs/concepts/security/pod-security-standards/) and determines whether to `admit` or `deny` the pod from running. Pod Security is the successor to [PodSecurityPolicy](/docs/concepts/policy/pod-security-policy/) which is deprecated in v1.21 and will be removed in v1.25. Let's explore how Pod Security works in the hope that cluster administrators and developers alike will use this to enforce secure defaults for their workloads. In this blog we will cover the key concepts of Pod Security along with how to use it. 

Reviewer 2: ```suggestion
With the release of Kubernetes v1.23, [Pod Security admission](/docs/concepts/security/pod-security-admission/) has now entered beta. Pod Security is a [built-in](/docs/reference/access-authn-authz/admission-controllers/) admission controller that evaluates pod specifications against a predefined set of [Pod Security Standards](/docs/concepts/security/pod-security-standards/) and determines whether to `admit` or `deny` the pod from running.

Pod Security is the successor to [PodSecurityPolicy](/docs/concepts/policy/pod-security-policy/) which was deprecated in the v1.21 release, and will be removed in Kubernetes v1.25. In this article, we will cover the key concepts of Pod Security along with how to use it. . We hope that cluster administrators and developers alike will use this new mechanism to enforce secure defaults for their workloads.
```

Could also hyperlink “specifications” to https://kubernetes.io/docs/concepts/overview/working-with-objects/kubernetes-objects/#object-spec-and-status if that sounds helpful.

## Why Pod Security

> 

Reviewer 2: ```suggestion

The overall aim of Pod Security is to let you isolate workloads. You can run a cluster that runs different workloads and, without adding extra third-party tooling, implement controls that require Pods for a workload to restrict their own privileges to a defined bounding set.

```

To me, the exact wording here isn't important, but a summary of what this new feature achieves is - separate from any mention of PodSecurityPolicy. Readers may never have encountered the PSP API, and may also not have read our previous blogs etc on the subject.

It's important to note that Pod Security doesn't have complete feature parity with the deprecated PodSecurityPolicy. Specifically, it doesn't have the ability to mutate or change Kubernetes resources to auto-remediate a policy violation on behalf of the user. Additionally, it doesn't provide fine-grained control over each allowed field and value within a pod specification or any other Kubernetes resource that you may wish to evaluate. If you need more fine-grained policy control then take a look at these [other](/docs/concepts/security/pod-security-standards/#faq) projects which support such use cases.

Pod Security also adheres to Kubernetes best practices of declarative object management by denying resources that violate the policy. This requires resources to be updated in source repositories, and tooling to be updated prior to being deployed to Kubernetes.

## How Does Pod Security Work?
    

> Pod Security is a built-in [admission controller](/docs/reference/access-authn-authz/admission-controllers/) starting with Kubernetes v1.22, but can also be run as a standalone [webhook](/docs/reference/access-authn-authz/extensible-admission-controllers/). Admission controllers function by intercepting requests in the Kubernetes API server prior to persistence to storage. They can either `admit` or `deny` a request. In the case of Pod Security, pod specifications will be evaluated against a configured policy in the form of a Pod Security Standard. This means that security sensitive fields in a pod specification will only be allowed to have [specific](h/docs/concepts/security/pod-security-standards/#profile-details) values. 

Reviewer 2: Is there a better place for “webhook” to hyperlink to / will there be, fairly soon?

Reviewer 5: https://github.com/kubernetes/website/blob/dev-1.23/content/en/docs/concepts/security/pod-security-admission.md#alternative-installing-the-podsecurity-admission-webhook is in the 1.23 branch (the anchor isn't very well-named though)

```

### Configure Pod Security

Policies are applied to a namespace via labels. These labels are as follows:
   * `pod-security.kubernetes.io/<MODE>: <LEVEL>` (required to enable pod security)
   * `pod-security.kubernetes.io/<MODE>-version: <VERSION>` (*optional*, defaults to latest)

A specific version can be supplied for each enforcement mode. The version pins the policy to the version that was shipped as part of the Kubernetes release. Pinning to a specific Kubernetes version allows for deterministic policy behavior while allowing flexibility for future updates to Pod Security Standards. The possible <MODE(S)> are `enforce`, `audit` and `warn`.

### When to use `warn`?

> The typical uses for `warn` are:

Reviewer 2: ```suggestion
The typical uses for `warn` are to get ready for a future change where you want to enforce
a different policy. The most two common cases would be:
```
How about case 3, you're not enforcing any policy yet, and you want to preview the impact? I think that's a good use for `warn` (maybe also: delete all existing Pods once you do).

I should add that I haven't actually tried Pod Security admission in any real-world cluster with a real world workload on it, so please take that in mind when considering my suggestions.

Reviewer 5: > How about case 3, you're not enforcing any policy yet, and you want to preview the impact? I think that's a good use for warn

That's sort of the same as "`warn` at a stricter level", where `enforce` is privileged and `warn` is baseline or restricted. If you think it's clearer to explicitly call that out as a third case, that's fine

> (maybe also: delete all existing Pods once you do).

If you're looking to check a new policy non-disruptively, deleting existing pods is not the way to do it :-)

* `warn` at the same level but a different version (e.g. pin `enforce` to *restricted+v1.23* and `warn` at *restricted+latest*)
* `warn` at a stricter level (e.g. `enforce` baseline, `warn` restricted)

It's not recommended to use `warn` for the exact same level+version of the policy as `enforce`. In the admission sequence, if `enforce` fails, the entire sequence fails before evaluating the `warn`.

First, create a namespace called `verify-pod-security` if not created earlier. For the demo, `--overwrite` is used when labeling to allow repurposing a single namespace for multiple examples.

```shell
kubectl create namespace verify-pod-security
```

### Deploy demo workloads

Each workload represents a higher level of security that would not pass the profile that comes after it.

> For the following examples, use the `busybox` container runs a `sleep` command for 1 million seconds (~11 days) or until deleted. Pod Security is not interested in the container, but rather the Kubernetes manifests and corresponding security values that are being admitted.

Reviewer 2: nits:
```suggestion
For the following examples, use the `busybox` container runs a `sleep` command for 1 million seconds (≅11 days) or until deleted. Pod Security is not interested in which container image you chose, but rather the Pod- and container- level settings and their implications for security. (There are third-party solutions that _do_ scan container images, and these also help with overall information security).
```
